### ISSUE TITLE
Undefined behaviour in winmm backend

### ISSUE THREAD

**Issue by barafael** [Boddlnagg/midir#137]
Here, a packet is filled by bytes from a message:

https://github.com/Boddlnagg/midir/blob/db24d55a873d0213e1c61329a2f84c886430a9ce/src/backend/winmm/mod.rs#L509-L513

This is undefined behaviour, because the address of an immutable `u32` is taken and cast to a `*mut u8`, which is then written to in an `unsafe` block.

I have fixes ready in #136 but these are part of a rather large patchset, so if you like I can send a small patch ahead, too.

**Comment by barafael** [comment]
Here's a playground which you can use with Miri to see that it's actually undefined behaviour: [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=use+std%3A%3Aptr%3A%3A%7Baddr_of%2C+addr_of_mut%7D%3B%0A%0Aconst+N%3A+usize+%3D+10%3B%0Aconst+DATA%3A+%5Bu8%3B+12%5D+%3D+%5B1%2C+2%2C+3%2C+4%2C+5%2C+6%2C+7%2C+8%2C+9%2C+10%2C+1%2C+212%5D%3B%0A%0Afn+main2%28%29+%7B%0A++++let+mut+packet%3A+u32+%3D+0%3B%0A++++let+ptr+%3D+addr_of_mut%21%28packet%29.cast%3A%3A%3Cu8%3E%28%29%3B%0A++++for+%28i%2C+item%29+in+DATA.iter%28%29.enumerate%28%29.take%28N%29+%7B%0A++++++++unsafe+%7B+*ptr.add%28i%29+%3D+*item+%7D%3B%0A++++%7D%0A++++println%21%28%22%7Bpacket%3Ab%7D%22%29%3B%0A%7D%0A%0Afn+main%28%29+%7B%0A++++let+mut+packet%3A+u32+%3D+0%3B%0A++++let+ptr+%3D+%26mut+packet+as+*mut+u32+as+*mut+u8%3B%0A++++for+i+in+0..N+%7B%0A++++++++unsafe+%7B+*ptr.offset%28i+as+isize%29+%3D+DATA%5Bi%5D+%7D%3B%0A++++%7D%0A++++println%21%28%22%7Bpacket%3Ab%7D%22%29%3B%0A%7D%0A)

**Comment by Boddlnagg** [comment]
Sorry, I have somehow missed this PR! I will probably just merge #136 eventually (after #135 and a rebase), I don't think a separate PR is necessary.

**Comment by barafael** [comment]
OK. I just did this PR because I wasn't sure you wanted to have the large blob of changes in #136 .
